ocw/lib/azure.py

```diff
@@ -172,10 +177,38 @@ def cleanup_disks_from_rg(self) -> None:
         for item in self.list_disks_by_resource_group(self.__resource_group):
             if self.is_outdated(item.changed_time):
                 if self.compute_mgmt_client().disks.get(self.__resource_group, item.name).managed_by:
-                    self.log_warn(f"Disk is in use - unable delete {item.name}")
+                    self.log_warn(f"Disk is in use - skipping {item.name}")
                 else:
                     if self.dry_run:
                         self.log_info(f"Deletion of disk {item.name} skipped due to dry run mode")
                     else:
                         self.log_info(f"Delete disk '{item.name}'")
                         self.compute_mgmt_client().disks.begin_delete(self.__resource_group, item.name)
+
+    def cleanup_galleries_from_rg(self) -> None:
+        self.log_dbg("Call cleanup_galleries_from_rg")
+
+        # Get ID's for all virtual machine scale sets and virtual machines
+        scale_set_gallery_ids = {
+            ss.virtual_machine_profile.storage_profile.image_reference.id.lower()
+            for ss in self.compute_mgmt_client().virtual_machine_scale_sets.list_all()
+            if ss.virtual_machine_profile.storage_profile.image_reference.id is not None
+        }
+        vm_gallery_ids = {
+            vm.storage_profile.image_reference.id.lower()
+            for vm in self.compute_mgmt_client().virtual_machines.list_all()
+            if vm.storage_profile.image_reference.id is not None
+        }
+        active_ids = scale_set_gallery_ids | vm_gallery_ids
+
+        for item in self.list_galleries_by_resource_group(self.__resource_group):
+            if self.is_outdated(item.changed_time):
+                gallery_id = item.id.lower()
+                if any(i.startswith(gallery_id) for i in active_ids):
```

Comment:
`active_ids` is dict , right ? so maybe something like `gallery_id in active_ids.keys()` :thinking:

Reply:
> `active_ids` is dict , right ? so maybe something like `gallery_id in active_ids.keys()` 🤔

It's a set.

Reply:
> ok than even simpler `gallery_id in active_ids` , why do we need `start_with` ? is there could be case that ID not fully match but we still don't want to treat it as the same ?

I with it were that simpler :)

Omitting the subscription in paths for `xxx`:

```
>>> client = ComputeManagementClient(credentials, subscription)
>>> mgmt_client = ResourceManagementClient(credentials, subscription)
>>> galleries = list(mgmt_client.resources.list_by_resource_group("openqa-upload", filter="resourceType eq 'Microsoft.Compute/galleries'", expand="changedTime"))
>>> galleries[0].id
'/subscriptions/xxx/resourceGroups/openqa-upload/providers/Microsoft.Compute/galleries/test_image_gallery'
>>> vms = list(client.virtual_machines.list_all())
>>> vm_gallery_ids = [vm.storage_profile.image_reference.id for vm in vms if vm.storage_profile.image_reference.id is not None]
>>> vm_gallery_ids[0]
'/subscriptions/xxx/resourceGroups/openqa-upload/providers/Microsoft.Compute/galleries/test_image_gallery/images/SLE-MICRO-5.2-BYOS-AZURE-X64-GEN2/versions/1.304.101'
```

So we have that `vm_gallery_ids[0]` starts with `galleries[0].id`.  Now I noticed that it's much safer to append a slash to `galleries[0].id` in the call to `startswith()`